lib/unison-sqlite/src/Unison/Sqlite/Connection.hs

```diff
@@ -0,0 +1,481 @@
+          params = Just params,
+          sql = s
+        }
+
+-- Without results, without parameters
+
+execute_ :: Connection -> Sql -> IO ()
+execute_ conn@(Connection _ _ conn0) s =
+  Sqlite.execute_ conn0 (coerce s) `catch` \(exception :: Sqlite.SQLError) ->
+    throwSqliteException
+      SqliteExceptionInfo
+        { connection = conn,
+          exception = SomeShowTypeable exception,
+          params = Nothing,
+          sql = s
+        }
+
+-- With results, with parameters, without checks
+
+queryList :: (Sqlite.FromRow b, Sqlite.ToRow a) => Connection -> Sql -> a -> IO [b]
+queryList conn@(Connection _ _ conn0) s params =
+  Sqlite.query conn0 (coerce s) params `catch` \(exception :: Sqlite.SQLError) ->
+    throwSqliteException
+      SqliteExceptionInfo
+        { connection = conn,
+          exception = SomeShowTypeable exception,
+          params = Just params,
+          sql = s
+        }
+
+queryListOne :: forall a b. (Sqlite.FromField b, Sqlite.ToRow a) => Connection -> Sql -> a -> IO [b]
+queryListOne conn s params =
+  coerce @(IO [Sqlite.Only b]) @(IO [b]) (queryList conn s params)
+
+queryMaybe :: (Sqlite.ToRow a, Sqlite.FromRow b) => Connection -> Sql -> a -> IO (Maybe b)
+queryMaybe conn s params =
+  queryListCheck conn s params \case
+    [] -> Right Nothing
+    [x] -> Right (Just x)
+    xs -> Left (ExpectedAtMostOneRowException (anythingToString xs))
+
+queryMaybeOne :: forall a b. (Sqlite.ToRow a, Sqlite.FromField b) => Connection -> Sql -> a -> IO (Maybe b)
+queryMaybeOne conn s params =
+  coerce @(IO (Maybe (Sqlite.Only b))) @(IO (Maybe b)) (queryMaybe conn s params)
+
+queryOne :: (Sqlite.FromRow b, Sqlite.ToRow a) => Connection -> Sql -> a -> IO b
+queryOne conn s params =
+  queryListCheck conn s params \case
+    [x] -> Right x
+    xs -> Left (ExpectedExactlyOneRowException (anythingToString xs))
+
+queryOneOne :: forall a b. (Sqlite.FromField b, Sqlite.ToRow a) => Connection -> Sql -> a -> IO b
+queryOneOne conn s params = do
+  coerce @(IO (Sqlite.Only b)) @(IO b) (queryOne conn s params)
+
+-- With results, with parameters, with checks
+
+queryListCheck ::
+  (Sqlite.FromRow b, Sqlite.ToRow a, Show e, Typeable e) =>
+  Connection ->
+  Sql ->
+  a ->
+  ([b] -> Either e r) ->
+  IO r
+queryListCheck conn s params check =
+  gqueryListCheck conn s params (mapLeft SomeShowTypeable . check)
+
+gqueryListCheck ::
+  (Sqlite.FromRow b, Sqlite.ToRow a) =>
+  Connection ->
+  Sql ->
+  a ->
+  ([b] -> Either SomeShowTypeable r) ->
+  IO r
+gqueryListCheck conn s params check = do
+  xs <- queryList conn s params
+  case check xs of
+    Left exception ->
+      throwSqliteException
+        SqliteExceptionInfo
+          { connection = conn,
+            exception,
+            params = Just params,
+            sql = s
+          }
+    Right result -> pure result
+
+queryListOneCheck ::
+  forall a b e r.
+  (Sqlite.FromField b, Sqlite.ToRow a, Show e, Typeable e) =>
```

Comment:
I think `Show e, Typeable e` might be too permissive of a constraint here. Almost any data satisfies this constraint so it may lead to mistakes where, after a refactor, something type-checks, but the `Left` value of the check hasn't been thought through.

I see that you have other reasons for not using `Exception` for this, but even a nominal typeclass

```
class (Show e, Typeable e) => SqliteException e
```

would at least allow us to search for everything that _could_ be thrown in a Check, and would prevent folks from (either on purpose or accidentally) throwing things like `()` as a Check exception.

Reply:
I think that's a reasonable suggestion. It will require more ceremony (one-line deriving of that type class for things you intend to use as exception context) but probably worth it.

Reply:
Since the main exception record is called `SqliteException`, how about `SqliteExceptionReason` for this type class?